persist/encoding/msgpack/decoder.go

```diff
@@ -0,0 +1,369 @@
+	if !ok {
+		return 0
+	}
+	actualType := dec.decodeObjectType()
+	if dec.err != nil {
+		return 0
+	}
+	if expectedType != actualType {
+		dec.err = fmt.Errorf("object type mismatch: expected %v actual %v", expectedType, actualType)
+		return 0
+	}
+	return numFieldsToSkip
+}
+
+func (dec *decoder) checkVersion(expected int) {
+	version := int(dec.decodeVarint())
+	if dec.err != nil {
+		return
+	}
+	if version > expected {
+		dec.err = fmt.Errorf("version mismatch: expected %v actual %v", expected, version)
+	}
+}
+
+func (dec *decoder) checkNumFieldsFor(objType objectType) (int, bool) {
+	actual := dec.decodeNumObjectFields()
+	if dec.err != nil {
+		return 0, false
+	}
+	expected := numFieldsForType(objType)
+	if expected > actual {
+		dec.err = fmt.Errorf("number of fields mismatch: expected %d actual %d", expected, actual)
+		return 0, false
+	}
+	return actual - expected, true
+}
+
+func (dec *decoder) skip(numFields int) {
+	if dec.err != nil {
+		return
+	}
+	if numFields < 0 {
+		dec.err = fmt.Errorf("number of fields to skip is %d", numFields)
+		return
+	}
+	for i := 0; i < numFields; i++ {
+		if err := dec.dec.Skip(); err != nil {
+			dec.err = err
+			return
+		}
+	}
+}
+
+func (dec *decoder) decodeNumObjectFields() int {
+	return int(dec.decodeArrayLen())
+}
+
+func (dec *decoder) decodeObjectType() objectType {
+	return objectType(dec.decodeVarint())
+}
+
+func (dec *decoder) decodeVarint() int64 {
+	if dec.err != nil {
+		return 0
+	}
+	value, err := dec.dec.DecodeInt64()
+	dec.err = err
+	return value
+}
+
+func (dec *decoder) decodeVarUint() uint64 {
+	if dec.err != nil {
+		return 0
+	}
+	value, err := dec.dec.DecodeUint64()
+	dec.err = err
+	return value
+}
+
+func (dec *decoder) decodeFloat64() float64 {
+	if dec.err != nil {
+		return 0.0
+	}
+	value, err := dec.dec.DecodeFloat64()
+	dec.err = err
+	return value
+}
+
+func (dec *decoder) decodeBytes() []byte {
+	if dec.err != nil {
+		return nil
+	}
+	// If we need to allocate new space for decoded byte slice, we delegate it to msgpack
+	// API which allocates a new slice under the hood, otherwise we simply locate the byte
+	// slice as part of the encoded byte stream and return it
+	var value []byte
+	if dec.newAllocForBytes {
+		value, dec.err = dec.dec.DecodeBytes()
+	} else {
+		bytesLen := dec.decodeBytesLen()
+		if dec.err != nil {
+			return nil
+		}
+		// NB(xichen): DecodeBytesLen() returns -1 if the byte slice is nil
+		if bytesLen == -1 {
+			return nil
+		}
+		numRead := len(dec.data) - dec.reader.Len()
+		value = dec.readBytes(numRead, bytesLen)
+	}
+	return value
+}
+
+func (dec *decoder) decodeArrayLen() int {
+	if dec.err != nil {
+		return 0
+	}
+	value, err := dec.dec.DecodeArrayLen()
+	dec.err = err
+	return value
+}
+
+func (dec *decoder) decodeBytesLen() int {
+	if dec.err != nil {
+		return 0
+	}
+	value, err := dec.dec.DecodeBytesLen()
+	dec.err = err
+	return value
+}
+
+func (dec *decoder) readBytes(start int, n int) []byte {
```

Comment:
May it be worth just inlining this whole function?  It's only used in one place from decodeBytes no?

Mainly because: it's strange that the parameters are absolute numbers, but the reader is moved a relative amount based on how much is being read from an absolute position.